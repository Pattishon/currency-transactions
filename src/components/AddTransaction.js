import React, { useState } from "react";
import { connect } from "react-redux";
import uuid from "uuid";
import { addTransaction } from "../store/actions/transactions";
import numeral from "numeral";
import "numeral/locales";

numeral.locale("pl");

export const AddTransaction = ({ addTransaction, converter }) => {
  const [error, setError] = useState("");
  const [description, setDescription] = useState("");
  const [amount, setAmount] = useState("");

  const handleDescriptionChange = e => {
    setDescription(e.target.value);
    setError("");
  };

  const handleAmountChange = e => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      setAmount(amount);
      setError("");
    } else if (amount.match(/^\d{1,}(,\d{0,2})?$/)) {
      const parsedAmount = amount.replace(",", ".");
      setAmount(parsedAmount);
      setError("");
    }
  };

  const handleAddTransaction = e => {
    e.preventDefault();
    if (!amount || !description) {
      setError("Proszę podać nazwę transakcji i wartość w Euro");
    } else {
      const newTransaction = {
        id: uuid(),
        description,
        amount: parseFloat(amount, 10) * 100
      };

      setError("");
      setDescription("");
      setAmount("");
      addTransaction(newTransaction);
    }
  };

  return (
    <div className="add-transaction">
      <h2 className="add-transaction__title">Dodaj nową transakcję</h2>
      <form onSubmit={handleAddTransaction}>
        {error ? <p className="add-transaction__error">{error}</p> : ""}
        <div>
          <label>
            Nazwa transakcji
            <input
              type="text"
              placeholder="nazwa transakcji"
              value={description}
              onChange={handleDescriptionChange}
              className="add-transaction__field"
            />
          </label>
        </div>
        <div>
          <label>
            Wartość w EURO:{" "}
            <input
              type="text"
              placeholder="wartość w EURO"
              value={amount}
              min="1"
              onChange={handleAmountChange}
              className="add-transaction__field"
            />
          </label>
        </div>
        <p>
          Wartość w PLN:{" "}
          {amount ? numeral(amount * converter).format("0.00") : "-"}
        </p>
        <button type="submit" className="button button--add">
          Dodaj transakcję
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = state => ({
  converter: state.converter,
  transactions: state.transactions
});

const mapDispatchToProps = dispatch => ({
  addTransaction: transaction => dispatch(addTransaction(transaction))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTransaction);
