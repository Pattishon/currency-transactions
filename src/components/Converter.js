import React, { useEffect } from "react";
import { connect } from "react-redux";
import { fetchState, updateConverter } from "../store/actions/transactions";
import numeral from "numeral";
import "numeral/locales";

numeral.locale("pl");

export const Converter = ({ converter, fetchState, updateConverter }) => {
  useEffect(() => {
    if (localStorage.converter) fetchState("converter");
  }, []);

  const handleConverterChange = e => {
    const value = e.target.value;

    if (!value || value.match(/^\d{1,}(\.\d{0,2})?$/)) {
      updateConverter(value);
    } else if (value.match(/^\d{1,}(,\d{0,2})?$/)) {
      const parsedVal = value.replace(",", ".");
      updateConverter(parsedVal);
    }
  };

  return (
    <div className="converter">
      1 EURO ={" "}
      <input
        type="text"
        placeholder="wpisz wartość w zł"
        value={converter}
        min={0}
        onChange={e => handleConverterChange(e)}
        className="converter__input"
      />{" "}
      PLN
    </div>
  );
};

const mapStateToProps = state => ({
  converter: state.converter
});
const mapDispatchToProps = dispatch => ({
  fetchState: key => dispatch(fetchState(key)),
  updateConverter: value => dispatch(updateConverter(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(Converter);
