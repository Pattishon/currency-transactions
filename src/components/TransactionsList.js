import React, { useEffect } from "react";
import { connect } from "react-redux";
import { deleteTransaction, fetchState } from "../store/actions/transactions";
import numeral from "numeral";
import "numeral/locales";

numeral.locale("pl");

export const TransactionsList = ({
  converter,
  fetchState,
  transactions,
  deleteTransaction
}) => {
  useEffect(() => {
    if (localStorage.transactions) fetchState("transactions");
  }, []);

  const handleDelete = id => {
    deleteTransaction(id);
  };

  return (
    <div className="transactions-list">
      <h2 className="transactions-list__title">Lista wszystkich transakcji</h2>
      {transactions.length > 0 ? (
        <ul className="transactions-list__list">
          {transactions.map(transaction => (
            <li key={transaction.id} className="transactions-list__list-item">
              <p>
                <span className="transactions-list__description">
                  {transaction.description}:{" "}
                </span>
                {numeral(transaction.amount / 100).format("0.00")} EURO (
                {numeral((transaction.amount * converter) / 100).format(
                  "0.00 $"
                )}
                )
              </p>
              <div>
                <button
                  onClick={() => handleDelete(transaction.id)}
                  className="button button--transactions-list"
                >
                  Usuń transakcję
                </button>
              </div>
            </li>
          ))}
        </ul>
      ) : (
        <p>Brak zapisanych transakcji.</p>
      )}
    </div>
  );
};

const mapStateToProps = state => ({
  converter: state.converter,
  transactions: state.transactions
});
const mapDispatchToProps = dispatch => ({
  deleteTransaction: id => dispatch(deleteTransaction(id)),
  fetchState: key => dispatch(fetchState(key))
});
export default connect(mapStateToProps, mapDispatchToProps)(TransactionsList);
