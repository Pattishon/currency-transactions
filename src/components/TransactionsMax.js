import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import "numeral/locales";

numeral.locale("pl");

export const TransactionsMax = ({ converter, transactions }) => {
  const max = transactions.reduce(
    (prev, current) => (prev.amount > current.amount ? prev : current),
    ""
  );
  const maxEuro = numeral(max.amount / 100).format("0.00");
  const maxPln = numeral((max.amount * converter) / 100).format("0.00 $");
  return (
    <div className="transactions-max">
      <h2 className="transactions-max__title">
        {" "}
        Transakcja o najwiÄ™kszej kwocie:
      </h2>
      <p className="transactions-max__transaction">
        {max
          ? `${max.description}: ${maxEuro} EURO (${maxPln}) `
          : "nie dodano jeszcze transakcji"}
      </p>
    </div>
  );
};

const mapStateToProps = state => ({
  converter: state.converter,
  transactions: state.transactions
});

export default connect(mapStateToProps, null)(TransactionsMax);
