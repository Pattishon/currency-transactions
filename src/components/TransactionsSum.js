import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import "numeral/locales";

numeral.locale("pl");

export const TransactionsSum = ({ converter, transactions }) => {
  const total = transactions.reduce((a, b) => a + (b.amount || 0), 0);
  const totalEuro = numeral(total / 100).format("0.00");
  const totalPln = numeral((total * converter) / 100).format("0.00 $");
  return (
    <div className="transactions-sum">
      <h2 className="transactions-sum-title">Suma wszystkich transakcji:</h2>{" "}
      {totalEuro} EURO ({totalPln})
    </div>
  );
};

const mapStateToProps = state => ({
  converter: state.converter,
  transactions: state.transactions
});

export default connect(mapStateToProps, null)(TransactionsSum);
