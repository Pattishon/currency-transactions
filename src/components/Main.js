import React from "react";
import AddTransaction from "./AddTransaction";
import Converter from "./Converter";
import TransactionsList from "./TransactionsList";
import TransactionsSum from "./TransactionsSum";
import TransactionsMax from "./TransactionsMax";

const Main = () => {
  return (
    <div className="app">
      <h1 className="title">Transakcje walutowe</h1>
      <div className="transactions">
        <div>
          <Converter />
          <AddTransaction />
          <TransactionsList />
          <TransactionsSum />
        </div>
        <TransactionsMax />
      </div>
    </div>
  );
};

export default Main;
