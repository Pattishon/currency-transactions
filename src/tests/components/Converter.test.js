import React from "react";
import { shallow } from "enzyme";
import { Converter } from "../../components/Converter";

let wrapper, fetchState, updateConverter;
beforeEach(() => {
  fetchState = jest.fn();
  updateConverter = jest.fn();
  wrapper = shallow(
    <Converter
      fetchState={fetchState}
      converter={2}
      updateConverter={updateConverter}
    />
  );
});

test("should render Converter correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should set converter on input change", () => {
  const value = "5";
  wrapper
    .find("input")
    .at(0)
    .simulate("change", {
      target: { value }
    });
  expect(updateConverter).toHaveBeenLastCalledWith(value);
});
