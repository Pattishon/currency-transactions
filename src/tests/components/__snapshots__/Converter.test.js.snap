// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Converter correctly 1`] = `
<div
  className="converter"
>
  1 EURO =
   
  <input
    className="converter__input"
    min={0}
    onChange={[Function]}
    placeholder="wpisz wartość w zł"
    type="text"
    value={2}
  />
   
  PLN
</div>
`;
